File: backend/internal/models/heat_map.go
package models

import "time"

// HeatMapPoint represents a geographic point with climbing activity
type HeatMapPoint struct {
	MPAreaID       int64     `json:"mp_area_id"`
	Name           string    `json:"name"`
	Latitude       float64   `json:"latitude"`
	Longitude      float64   `json:"longitude"`
	ActivityScore  int       `json:"activity_score"` // Weighted: recent=higher
	TotalTicks     int       `json:"total_ticks"`
	ActiveRoutes   int       `json:"active_routes"`
	LastActivity   time.Time `json:"last_activity"`
	UniqueClimbers int       `json:"unique_climbers"`
	HasSubareas    bool      `json:"has_subareas"`
}

// AreaActivityDetail provides comprehensive activity data for an area
// Reuses existing ClimbHistoryEntry, CommentSummary, and RouteActivitySummary types
type AreaActivityDetail struct {
	MPAreaID         int64             `json:"mp_area_id"`
	Name             string            `json:"name"`
	ParentMPAreaID   *int64            `json:"parent_mp_area_id,omitempty"`
	Latitude         *float64          `json:"latitude,omitempty"`
	Longitude        *float64          `json:"longitude,omitempty"`
	TotalTicks       int               `json:"total_ticks"`
	ActiveRoutes     int               `json:"active_routes"`
	UniqueClimbers   int               `json:"unique_climbers"`
	LastActivity     time.Time         `json:"last_activity"`
	RecentTicks      []TickDetail      `json:"recent_ticks"`
	RecentComments   []CommentSummary  `json:"recent_comments"`
	ActivityTimeline []DailyActivity   `json:"activity_timeline"`
	TopRoutes        []TopRouteSummary `json:"top_routes"`
}

// TickDetail represents a single tick for area detail views
type TickDetail struct {
	MPRouteID int64     `json:"mp_route_id"`
	RouteName string    `json:"route_name"`
	Rating    string    `json:"rating"`
	UserName  string    `json:"user_name"`
	ClimbedAt time.Time `json:"climbed_at"`
	Style     string    `json:"style"`
	Comment   string    `json:"comment"`
}

// CommentSummary represents a comment with basic info
type CommentSummary struct {
	ID          int64     `json:"id"`
	UserName    string    `json:"user_name"`
	CommentText string    `json:"comment_text"`
	CommentedAt time.Time `json:"commented_at"`
	MPRouteID   *int64    `json:"mp_route_id,omitempty"`
	RouteName   *string   `json:"route_name,omitempty"`
}

// DailyActivity represents activity count for a single day
type DailyActivity struct {
	Date       time.Time `json:"date"`
	TickCount  int       `json:"tick_count"`
	RouteCount int       `json:"route_count"`
}

// RouteActivity represents a route with location and activity
type RouteActivity struct {
	MPRouteID    int64     `json:"mp_route_id"`
	Name         string    `json:"name"`
	Rating       string    `json:"rating"`
	Latitude     *float64  `json:"latitude"`
	Longitude    *float64  `json:"longitude"`
	TickCount    int       `json:"tick_count"`
	LastActivity time.Time `json:"last_activity"`
	MPAreaID     int64     `json:"mp_area_id"`
	AreaName     string    `json:"area_name"`
}

// TopRouteSummary is a simplified route with activity count for top routes lists
type TopRouteSummary struct {
	MPRouteID    int64     `json:"mp_route_id"`
	Name         string    `json:"name"`
	Rating       string    `json:"rating"`
	TickCount    int       `json:"tick_count"`
	LastActivity time.Time `json:"last_activity"`
}

File: frontend/src/types/heatmap.ts
// Heat Map Types for Climbing Activity Visualization

export interface HeatMapPoint {
  mp_area_id: number;
  name: string;
  latitude: number;
  longitude: number;
  activity_score: number;
  total_ticks: number;
  active_routes: number;
  last_activity: string; // ISO date string
  unique_climbers: number;
  has_subareas: boolean;
}

export interface TickDetail {
  mp_route_id: number;
  route_name: string;
  rating: string;
  user_name: string;
  climbed_at: string; // ISO date string
  style: string;
  comment: string;
}

export interface CommentSummary {
  id: number;
  user_name: string;
  comment_text: string;
  commented_at: string; // ISO date string
  mp_route_id?: number;
  route_name?: string;
}

export interface DailyActivity {
  date: string; // ISO date string
  tick_count: number;
  route_count: number;
}

export interface TopRouteSummary {
  mp_route_id: number;
  name: string;
  rating: string;
  tick_count: number;
  last_activity: string; // ISO date string
}

export interface AreaActivityDetail {
  mp_area_id: number;
  name: string;
  parent_mp_area_id?: number;
  latitude?: number;
  longitude?: number;
  total_ticks: number;
  active_routes: number;
  unique_climbers: number;
  last_activity: string; // ISO date string
  recent_ticks: TickDetail[];
  recent_comments: CommentSummary[];
  activity_timeline: DailyActivity[];
  top_routes: TopRouteSummary[];
}

export interface RouteActivity {
  mp_route_id: number;
  name: string;
  rating: string;
  latitude?: number;
  longitude?: number;
  tick_count: number;
  last_activity: string; // ISO date string
  mp_area_id: number;
  area_name: string;
}

export interface HeatMapActivityResponse {
  points: HeatMapPoint[];
  count: number;
  filters: {
    start_date: string;
    end_date: string;
    min_activity: number;
    limit: number;
  };
}

export interface RoutesResponse {
  routes: RouteActivity[];
  count: number;
}

export interface GeoBounds {
  minLat: number;
  maxLat: number;
  minLon: number;
  maxLon: number;
}
